### DISCUSSION TITLE
Request and Response in Kurtosis

### DISCUSSION THREAD

**Discussion by shreyasbhat0** [kurtosis-tech/kurtosis#684] [Q&A] [answered]
I need to make a request to a service, and wait until an array in the response body has a nonzero length. How can I do this in Kurtosis?

**Comment by mieubrisse** [comment, marked as answer]
This can be done via the `extract` field that's available on [GetHttpRequestRecipe](https://docs.kurtosis.com/starlark-reference/get-http-request-recipe), [PostHttpRequestRecipe](https://docs.kurtosis.com/starlark-reference/post-http-request-recipe), and [ExecRecipe](https://docs.kurtosis.com/starlark-reference/exec-recipe) in combination with [plan.wait](https://docs.kurtosis.com/starlark-reference/plan#wait).

The way `extract` works is that, when the Recipe is run (via [plan.request](https://docs.kurtosis.com/starlark-reference/plan/#request) for Get/PostHttpRecipe and [plan.exec](https://docs.kurtosis.com/starlark-reference/plan#exec) for ExecRecipe), the result object will be a dictionary containing [future references](https://docs.kurtosis.com/concepts-reference/future-references) to the extracted values.

E.g. if you define a recipe with an `extract` of `valueOfMyKey`, `response["extract.valueOfMyKey"]` will be a future to that value at execution time:

```starlark
recipe = GetHttpRequestRecipe(
    port_id = "my_port",
    endpoint = "/some-endpoint",
    extract = {
        # Uses jq to extract the value of the key
        "valueOfMyKey" : ".myKey",
    },
)
response = plan.request(service = "my-service", recipe = recipe)

# Contains a future reference to the value of "myKey"
plan.print(response["extract.valueOfMyKey"])
```

So in your case, to wait until the length of an array is non-zero, let's suppose that your HTTP request is returning a body like this:

```javascript
{
  "theArray": [],
}
```

We can get the length in `extract`, and then use it in a `wait`:

```starlark
recipe = recipe = GetHttpRequestRecipe(
    port_id = "my_port",
    endpoint = "/some-endpoint",
    extract = {
        # Leverage jq's length function
        "arrayLen" : ".theArray | length",
    },
)
final_response = plan.wait(
    service_name = "my-service",
    recipe = recipe,
    field = "extract.arrayLen",
    assertion = ">",
    target_value = 0,
)
plan.print("The array length after passing the wait is: " + final_response["extract.arrayLen"])
```
